py_linq/py_linq.py

```diff
@@ -141,6 +150,33 @@ def median(self, func=lambda x: x):
             else (float(result[i - 1]) + float(result[i])) / float(2)
         )
 
+    def index_of(self, element):
```

Comment:
The method signature for this should include some `lambda` function to act as a selector for more complicated objects that do not have an `__eq__` method defined (ie. how to define equality between 2 different instances of an object with same properties).

Also, what if your collection contains different types of objects (which is possible in an `Enumerable` collection). How do you define equality between these 2 different objects? This is what the `lambda` function would be for. As a default it could be set as `lambda x: x`

For example:

```
class Point(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y
```

```
points = Enumerable([Point(1, 2), Point(1, 3)])
assert points.index_of(Point(1,2)) is not None
```
This above example would fail in the assert because Point(1,2) are two different instances of Point even though they have the same x, y values.

Reply:
Good point, I guess originally I was just leaving it up to them to implement __eq__ or do a select _first_ then index_of to account for what's needed beforehand, but an optional let's call it normalization function sounds like a good idea. I'll update in a bit.

Reply:
Bit of an impasse on this, issue #72 (not specifically with regards to any() though) makes the behavior of this difficult to predict.

Here's what I wanted to do:

**Updated index_of function:**

```python
    def index_of(self, element, func=lambda x: x, apply_func_to_element=False):
        """
        Returns the index of the first occurrence of a given element.

        :param element: the element for which to retrieve the index
        :param func: optional selector to apply to the collection as lambda expression for equality comparison
        :param apply_func_to_element: optional boolean flag indicating whether the selector lambda should be applied to the passed in element as well - default is False
        :return: Index of given element
        """
        for i, e in enumerate(self):
            if func(e) == (func(element) if apply_func_to_element else element):
                return i

        return None
```

**Set up for testing:**
```python
class Rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width

    def __eq__(self, other):
        if isinstance(other, Rectangle):
            return self.length == other.length and self.width == other.width

    def area(self):
        return self.length * self.width

    def perimeter(self):
        return 2 * self.length + 2 * self.width


class Square(Rectangle):
    def __init__(self, length):
        super().__init__(length, length)

self.complex_types = Enumerable([
    Rectangle(1, 2),
    Rectangle(1, 4),
    Rectangle(2, 1),
    Rectangle(3, 4),
    Square(1),
    Square(2),
])
```

**Test:**

```python
    def test_index_of_complex_types(self):
        self.assertEqual(4, self.complex_types.index_of(Square(1)))
        self.assertEqual(1, self.complex_types.index_of(4, lambda x: x.area()))
        self.assertEqual(1, self.complex_types.index_of(Square(2), lambda x: x.area(), True)     
```

Those assertions would pass if they were the only ones being executed, but since the Enumerable's order gets changed after a call to any of them, index_of becomes unreliable thereafter.

```python
    def test_index_of_complex_types(self):
        # issue #72 will leave to undesirable behavior without an order_by happening first
        complex_types = self.complex_types.order_by(lambda x: (x.length, x.width))
        # -> [ Square(1), Rectangle(1,2), Rectangle(1,4), Rectangle(2,1), Square(2), Rectangle(3,4)]
        self.assertEqual(2, complex_types.index_of(4, lambda x: x.area()))

        # have to do it again unfortunately
        complex_types = self.complex_types.order_by(lambda x: (x.length, x.width))
        self.assertEqual(2, complex_types.index_of(Square(2), lambda x: x.area(), True))
```

What are your thoughts on this?